src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,168 @@
+            return snapshot.Storages[CreateStorageKey(Prefix_NodeList)].GetInteroperable<NodeList>().ToArray();
+        }
+
+        public OracleRequest GetRequest(StoreView snapshot, ulong id)
+        {
+            return snapshot.Storages.TryGet(CreateStorageKey(Prefix_Request).Add(id))?.GetInteroperable<OracleRequest>();
+        }
+
+        public IEnumerable<OracleRequest> GetRequests(StoreView snapshot)
+        {
+            return snapshot.Storages.Find(new byte[] { Prefix_Request }).Select(p => p.Value.GetInteroperable<OracleRequest>());
+        }
+
+        public IEnumerable<OracleRequest> GetRequestsByUrl(StoreView snapshot, string url)
+        {
+            IdList list = snapshot.Storages.TryGet(CreateStorageKey(Prefix_IdList).Add(GetUrlHash(url)))?.GetInteroperable<IdList>();
+            if (list is null) yield break;
+            foreach (ulong id in list)
+                yield return snapshot.Storages[CreateStorageKey(Prefix_Request).Add(id)].GetInteroperable<OracleRequest>();
+        }
+
+        private static byte[] GetUrlHash(string url)
+        {
+            return Crypto.Hash160(Utility.StrictUTF8.GetBytes(url));
+        }
+
+        internal override void Initialize(ApplicationEngine engine)
+        {
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_NodeList), new StorageItem(new NodeList()));
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RequestId), new StorageItem(BitConverter.GetBytes(0ul)));
+        }
+
+        protected override void PostPersist(ApplicationEngine engine)
+        {
+            base.PostPersist(engine);
+            foreach (Transaction tx in engine.Snapshot.PersistingBlock.Transactions)
+            {
+                //Filter the response transactions
+                OracleResponse response = tx.Attributes.OfType<OracleResponse>().FirstOrDefault();
+                if (response is null) continue;
+
+                //Remove the request from storage
+                StorageKey key = CreateStorageKey(Prefix_Request).Add(response.Id);
+                OracleRequest request = engine.Snapshot.Storages[key].GetInteroperable<OracleRequest>();
+                engine.Snapshot.Storages.Delete(key);
+
+                //Remove the id from IdList
+                key = CreateStorageKey(Prefix_IdList).Add(GetUrlHash(request.Url));
+                IdList list = engine.Snapshot.Storages.GetAndChange(key).GetInteroperable<IdList>();
+                if (!list.Remove(response.Id)) throw new InvalidOperationException();
+                if (list.Count == 0) engine.Snapshot.Storages.Delete(key);
+
+                //Mint GAS for oracle nodes
+                ECPoint[] nodes = GetOracleNodes(engine.Snapshot);
+                int index = (int)(response.Id % (ulong)nodes.Length);
+                UInt160 account = Contract.CreateSignatureRedeemScript(nodes[index]).ToScriptHash();
+                GAS.Mint(engine, account, OracleRequestPrice);
+            }
```

Comment:
We can refund gas here,if user paid an excess fee.

Reply:
If the sender is the hash of the oracle contract, there is no way to refund.

Reply:
We can add a Requester in Request.

Reply:
I know this can be solved, but is it worth making it so complicated?

Reply:
Replace txid directly when requesting

Reply:
Replacing txid, replacing the sender, I think they are the same.

Reply:
From the user's perspective, a non-refundable fee will cause them to pay excess fees, which is a loss.From the perspective of the code modification, it is not a big change.
When we discussed the synchronous and asynchronous solutions, we had listed their respective advantages and disadvantages.It is acceptable to me if such a level of change can solve these problems.

Reply:
Refunding from SystemFee is a bad precedent.

Reply:
Maybe we can use `Transfer` or `Burn` method.

Reply:
If you use `transfer`, you can't create new requests in callback.

Reply:
If we don't refund, user need to set a "correct" `gasForRepsonse`, which may be not convenient if consider the recursive nesting of response -> request.

Reply:
In fact, `gasForRepsonse` is set by the contract who calls the oracle contract.

Reply:
Got, the code looks good for me.